detect_secrets/core/secrets_collection.py

```diff
@@ -1,396 +1,240 @@
+    def __iter__(self) -> Generator[Tuple[str, PotentialSecret], None, None]:
+        for filename in sorted(self.files):
+            secrets = self[filename]
 
-    def _extract_secrets_from_file(self, f, filename):
-        """Extract secrets from a given file object.
+            # TODO: Handle cases when line numbers are not supplied
+            for secret in sorted(secrets, key=lambda x: (x.line_number, x.secret_hash)):
+                yield filename, secret
 
-        :type f:        File object
-        :type filename: string
+    def __bool__(self) -> bool:
+        # This checks whether there are secrets, rather than just empty files.
+        # Empty files can occur with SecretsCollection subtraction.
+        return bool(list(self))
+
+    def __eq__(self, other: Any, strict: bool = False) -> bool:
         """
-        try:
-            log.info('Checking file: %s', filename)
+        :param strict: if strict, will return False even if secrets match
+            (e.g. if line numbers are different)
+        """
+        if not isinstance(other, SecretsCollection):
+            raise NotImplementedError
 
-            for results, plugin in self._results_accumulator(filename):
-                results.update(plugin.analyze(f, filename))
-                f.seek(0)
+        if self.files != other.files:
+            return False
 
-        except UnicodeDecodeError:
-            log.warning('%s failed to load.', filename)
+        for filename in self.files:
+            self_mapping = {secret.secret_hash: secret for secret in self[filename]}
+            other_mapping = {secret.secret_hash: secret for secret in other[filename]}
 
-    def _extract_secrets_from_patch(self, f, plugin, filename):
-        """Extract secrets from a given patch file object.
+            # Since PotentialSecret is hashable, we compare their identities through this.
+            if set(self_mapping.values()) != set(other_mapping.values()):
+                return False
 
-        Note that we only want to capture incoming secrets (so added lines).
-        Note that this is only called by detect-secrets-server.
+            if not strict:
+                continue
 
-        :type f: unidiff.patch.PatchedFile
-        :type plugin: detect_secrets.plugins.base.BasePlugin
-        :type filename: str
-        """
-        output = {}
-        for chunk in f:
-            # target_lines refers to incoming (new) changes
-            for line in chunk.target_lines():
-                if line.is_added:
-                    output.update(
-                        plugin.analyze_line(
-                            line.value,
-                            line.target_line_no,
-                            filename,
-                        ),
-                    )
+            for secretA in self_mapping.values():
+                secretB = other_mapping[secretA.secret_hash]
 
-        return output
+                valuesA = vars(secretA)
```

Comment:
We actually already do this on the former line:

```python
if set(self_mapping.values()) != set(other_mapping.values()):
    return False
```

Since the `__hash__` function leverages `fields_to_compare`, set comparison performs this check.

This logic handles "strict" comparison, which compares things like line numbers, verification status and labelled status, which is why we check for things other than `fields_to_compare`.